src/python/zquantum/core/cost_function_test.py

```diff
@@ -22,3 +22,22 @@ def test_ansatz_based_cost_function_returns_value_between_plus_and_minus_one(
     params = np.array([1])
     value = ansatz_based_cost_function(params)
     assert -1 <= value <= 1
+
+
+@pytest.fixture
+def noisy_ansatz():
+    target_operator = QubitOperator("Z0")
+    ansatz = MockAnsatz(number_of_layers=2, problem_size=1)
+    backend = MockQuantumSimulator()
+    estimator = MockEstimator()
+    return AnsatzBasedCostFunction(
+        target_operator, ansatz, backend, estimator=estimator, parameter_precision=1e-4
+    )
+
+
+def test_ansatz_based_cost_function_adds_noise_to_parameters(noisy_ansatz):
+    params = np.array([0.1, 2.3])
```

Comment:
The MockSimulator that is used in the test actually just returns random values so I don't think we can use it to check that the average of the cost function value corresponds to the correct value, if that is what you meant.

Reply:
There is a bigger problem with this test. Since MockSimulator used in the test returns random values, this test passes even if the ansatz used is not noisy.

I can see two possible solutions:
- we remove this test,
- we test whether actually `self.ansatz.get_executable_circuit` gets called with different parameters. This can be done by spying on `get_executable_circuit` by temporarily replacing it with a mock.

Let me know what you think would be better :slightly_smiling_face:

Reply:
Excellent point... I only recently realized MockSimulator returns random values and I didn't make the connection to the test being useless, so thank you :)
I like to have a test for a new functionality, but I'm not quite sure how to implement a mock and perform the spying. Would you have some pointers?

Reply:
Sorry for the delay, I'll show you an example of how to do this. Hopefully, I'll manage to do it today :)